server/src/main/java/com/flipkart/varadhi/services/SubscriptionService.java

```diff
@@ -49,89 +110,270 @@ public VaradhiSubscription createSubscription(
-            return controllerClient.stopSubscription(subscriptionName, requestedBy);
-        }
-        throw new InvalidOperationForResourceException(
-                "Subscription is in state %s. It can't be started/stopped.".formatted(
-                        subscription.getStatus().getState()));
-    }
-
-    private void validateCreation(VaradhiTopic topic, VaradhiSubscription subscription) {
-        if (subscription.isGrouped() && !topic.isGrouped()) {
-            throw new IllegalArgumentException(
-                    "Cannot create grouped Subscription as it's Topic(%s) is not grouped".formatted(
-                            subscription.getTopic()));
-        }
+        return performSubscriptionOperation(subscriptionName, requestedBy, controllerClient::stopSubscription);
     }
 
+    /**
+     * Updates an existing subscription.
+     *
+     * @param subscriptionName  the name of the subscription
+     * @param fromVersion       the current version of the subscription
+     * @param description       the new description of the subscription
+     * @param grouped           whether the subscription is grouped
+     * @param endpoint          the new endpoint of the subscription
+     * @param retryPolicy       the new retry policy of the subscription
+     * @param consumptionPolicy the new consumption policy of the subscription
+     * @param requestedBy       the user requesting the update
+     *
+     * @return a CompletableFuture representing the updated subscription
+     */
     public CompletableFuture<VaradhiSubscription> updateSubscription(
             String subscriptionName, int fromVersion, String description, boolean grouped, Endpoint endpoint,
             RetryPolicy retryPolicy, ConsumptionPolicy consumptionPolicy, String requestedBy
     ) {
-        VaradhiSubscription existingSubscription = metaStore.getSubscription(subscriptionName);
-        validateForConflictingUpdate(fromVersion, existingSubscription.getVersion());
-        VaradhiTopic subscribedTopic = metaStore.getTopic(existingSubscription.getTopic());
-        validateForSubscribedTopic(subscribedTopic, grouped);
-        return controllerClient.getSubscriptionState(subscriptionName, requestedBy).thenApply(ss -> {
-            existingSubscription.setGrouped(grouped);
-            existingSubscription.setDescription(description);
-            existingSubscription.setEndpoint(endpoint);
-            existingSubscription.setRetryPolicy(retryPolicy);
-            existingSubscription.setConsumptionPolicy(consumptionPolicy);
+        VaradhiSubscription subscription = getValidatedSubscription(subscriptionName);
+        validateVersionForUpdate(fromVersion, subscription.getVersion());
+
+        boolean originalGrouped = subscription.isGrouped();
+
+        subscription.setGrouped(grouped);
+        validateGroupedSubscription(metaStore.getTopic(subscription.getTopic()), subscription);
+
+        subscription.setGrouped(originalGrouped);
```

Comment:
why are we setting originalGrouped again?

Reply:
Modifying the returned value from the method is risky. 
We should clone it before modifying.

Lets follow the receiver of the object is supposed to make a copy.
For eg: If the method wants to modify the passed parameters, then the method (callee) is supposed to make the copy. 
If the caller wants to modify the returned value from the method, then the caller needs to make the copy.